DBoW2/DBoW2_sample_code.cpp

```diff
@@ -0,0 +1,218 @@
+#include <opencv2/core.hpp>
+#include <opencv2/highgui.hpp>
+#include <opencv2/features2d.hpp>
+#include <easy/profiler.h>
+
+#include <iostream>
+#include <vector>
+#include <filesystem>
+#include <string>
+#include <memory>
+
+using namespace DBoW2;
+namespace fs = std::filesystem;
+namespace UnitTest
+{
+  using FeatureVector = std::vector<std::vector<cv::Mat>>;
+  using Features = std::vector<cv::Mat>;
+  using FileLUT = std::vector<std::string>;
+}
+
+bool loadDBFeatures(UnitTest::FeatureVector &vfeatures, UnitTest::FileLUT &dbTable, const cv::Ptr<cv::ORB> &orb);
+bool loadVocabulary(OrbVocabulary &voc);
+void changeStructure(const cv::Mat &plain, UnitTest::Features &out);
+bool createDatabase(OrbDatabase &db, const UnitTest::FeatureVector &vfeatures);
+bool loadQueryFeatures(UnitTest::FeatureVector &vfeatures, const cv::Ptr<cv::ORB> &orb);
+bool queryDatabase(OrbDatabase &db, const UnitTest::FeatureVector &vfeatures, UnitTest::FileLUT &queryResultTable, const UnitTest::FileLUT &dbTable);
+bool saveResult(const UnitTest::FileLUT &queryResultTable);
+
+
+// Number of Images to Build Database
+constexpr int NDBIMAGES = 4;
+// Number of Images to Query with Database
+constexpr int NQUERYIMAGES = 1;
+
+
+// Paths 
+fs::path vocPath("../ORBvoc/ORBvoc.txt");
+fs::path dbPath("../demo/Database/");
+fs::path queryPath("../demo/Query/");
+fs::path savePath("../demo/Result/");
+
+
+int main()
+{
+  EASY_MAIN_THREAD;
+  EASY_PROFILER_ENABLE;
+
+  // DataStructures
+  UnitTest::FileLUT queryResultTable;
+  UnitTest::FileLUT dbTable;
+  UnitTest::FeatureVector dbFeatures;
+  UnitTest::FeatureVector queryFeatures;
+  cv::Ptr<cv::ORB> ORB = cv::ORB::create();
+  std::unique_ptr<OrbVocabulary> ptrVocabulary(new OrbVocabulary());
+
+  if(!loadDBFeatures(dbFeatures, dbTable, ORB))
+    std::cerr << "Error while loading DB features\n" << std::endl;
+
+  if(!loadVocabulary(*ptrVocabulary))
+    std::cerr << "Error while loading Vocabulary\n" << std::endl;
+
+  EASY_BLOCK("Initialize Database", profiler::colors::LightBlue);
+  std::unique_ptr<OrbDatabase> ptrDatabase(new OrbDatabase(*ptrVocabulary, false, 0));
+  EASY_END_BLOCK;
+
+  if(!createDatabase(*ptrDatabase, dbFeatures))
+    std::cerr << "Error while Creating DB\n" << std::endl;
+
+  if(!loadQueryFeatures(queryFeatures, ORB))
+    std::cerr << "Error loading Query features\n" << std::endl;
+
+  if(!queryDatabase(*ptrDatabase, queryFeatures, queryResultTable, dbTable))
+    std::cerr << "Error while Querying DB\n" << std::endl;
+
+  if(!saveResult(queryResultTable))
+    std::cerr << "Error while saving Results\n" << std::endl; 
+
+  profiler::dumpBlocksToFile("DBoW2.prof");
+  return 0;
+}
+
+
+bool loadDBFeatures(UnitTest::FeatureVector &vfeatures, UnitTest::FileLUT &dbTable, const cv::Ptr<cv::ORB> &orb)
+{
+  EASY_FUNCTION("Load DB Features", profiler::colors::LightGreen);
+
+  vfeatures.clear();
+  vfeatures.reserve(NDBIMAGES);
+
+  for(fs::directory_iterator it(dbPath); it != fs::end(it); it++)

```

Comment:
```suggestion
for (const auto& entry : dbPath)

```
뭔가 느낌상 이게 될 것 같다는 생각이 드는군요...!
Range-based for loop 방식인데, 이 방법 한번 사용해보시겠어요? 안돌아가면 수정 안하셔도 됩니다

Reply:
dbPath를 넣어주면 컴파일 에러가 나서 
`fs::directory_iterator it(dbPath);
 for(const auto& entry : it) `
이런 식으로 작성했는데 괜찮을까요?

Reply:
바로 dbPath를 넣어주면 컴파일 에러가 나서 
`for(const auto& entry : fs::directory_iterator(dbPath)`
이런식으로 작성했는데 괜찮을까요?